### PR TITLE
Added option to upload a new format to a given book

### PR DESCRIPTION
**Pull request by Kyosfonica**
Implements [219](https://github.com/janeczku/calibre-web/issues/219)

From the edit metadata form:
Added option to upload a file in the form so there is the option to have different formats of the same book.
Uploadable formats are the same as for new books entries.

### CODE REVIEW

**Comment by jkrehm** [comment]
If you rebase onto `master` you should be able to clear out the conflict easily. Then force push to update your PR.

File: cps/web.py [hunk 1]
```diff
@@ -2719,6 +2719,35 @@ def edit_book(book_id):
...
+        file_ext = requested_file.filename.rsplit('.', 1)[-1].lower()
+        if file_ext not in ALLOWED_EXTENSIONS:
+            flash(_('File extension "%s" is not allowed to be uploaded to this server' % file_ext), category="error")
+            return redirect(url_for('index'))
+
+        file_name = book.path.rsplit('/', 1)[-1]
+        filepath = config.config_calibre_dir + os.sep + book.path
+        filepath = os.path.normpath(filepath)
+        saved_filename = filepath + os.sep + file_name + '.' + file_ext
+        file_size = os.path.getsize(requested_file.stream.name)
```
Comment (jkrehm) [inline cps/web.py]: This line caused a problem for me. When I moved it to be right before line 2743 and changed it to `file_size = os.path.getsize(saved_filename)` it worked correctly.

**Comment by OzzieIsaacs** [comment]
Hi,

this looks like a promising PR. But I'm not 100% happy. If you select a 2nd file there is no feedback that which file is selected:
upload_invis
I changed the visibility of the file-button, so you can see that something is seleted:
visi

Could you please add this to your code (except the ugly button). If I understood it right than is the only posibility to add additional formats during upload of the book. Qould it be easy to add this function also to the "Edit Metadata"-page to have this function always available (it's not a precondition to have this before merge, just a question)

**Comment by Kyosfonica** [comment]
I'll recheck the code. On chrome the filename was visible. this button is available both when uploading a new book and when editing metadata.

**Comment by OzzieIsaacs** [comment]
Okay. On Firefox the text isn't visible as the screenshots show (or don't show)

**Comment by Kyosfonica** [comment]
Reviewed on Firefox. It seems like a JS error that prevents the full js code to run.
Chrome:
image
No JS errors.

Firefox:
image
JS error at line 26 in edit_book.js.
That line, if I'm not wrong, the one that writes 'dd/mm/aaaa' inside the publishing date field.
`$("#fake_pubdate")
                .val(pubDate.toLocaleDateString(language))
                .removeClass("hidden");`

**Comment by Kyosfonica** [comment]
Please, recheck this PR as I rebased the code of the branch to match master and I am not sure if I done this right. I see a lot of new files that I didn't modified.

**Comment by Kyosfonica** [comment]
Please, recheck this PR as I rebased the code of the branch and I am not sure if I done this right.
In case of error I created a brand new branch with the latest update of master and just my changes.

I want to learn how to deal with this kind of situations where the master branch of the main project updates and my changes in a branched fork have to catch up.

File: cps/web.py (hunk 1 above)
Comment (Kyosfonica) [reply cps/web.py]: Cool, moved the line to prevent the error.

**Pull request merged**